src/AsyncProducers.cpp

```diff
@@ -692,40 +690,61 @@ class UpdateIndices : public IRMutator {
     }
 };
 
-// Inject double buffering.
-class InjectDoubleBuffering : public IRMutator {
+// Inject ring buffering.
+class InjectRingBuffering : public IRMutator {
     using IRMutator::visit;
 
+    struct Loop {
+        std::string name;
+        Expr min;
+        Expr extent;
+
+        Loop(std::string n, Expr m, Expr e)
+            : name(std::move(n)), min(std::move(m)), extent(std::move(e)) {
+        }
+    };
+
+    const map<string, Function> &env;
+    std::vector<Loop> loops;
+    std::map<std::string, int> hoist_storage_loop_index;
+
     Stmt visit(const Realize *op) override {
         Stmt body = mutate(op->body);
         Function f = env.find(op->name)->second;
         Region bounds = op->bounds;
         if (f.schedule().ring_buffer().defined()) {
-            std::string enclosing_loop_var = loop_names.back();
-
+            // For the ring buffering we expand the storage by adding another dimension of
+            // the range of [0, ring_buffer.extent].
             bounds.emplace_back(0, f.schedule().ring_buffer());
-            Expr current_index = Load::make(Int(32), f.name() + ".ring_buffer.index", 0, Buffer<>(), Parameter(), const_true(), ModulusRemainder());
+            // Build an index for accessing ring buffer as a linear combination of all
+            // loop variables between the storage location (defined by the HoistStorage loop level)
+            // and corresponding Realize node.
+            int loop_index = hoist_storage_loop_index[op->name] + 1;
+            Expr current_index = Variable::make(Int(32), loops[loop_index].name);
+            while (++loop_index < (int)loops.size()) {
+                current_index = current_index *
+                                    (loops[loop_index].extent - loops[loop_index].min) +
+                                Variable::make(Int(32), loops[loop_index].name);
+            }
+            current_index = current_index % f.schedule().ring_buffer();
```

Comment:
Is this guaranteed to be > 0?

Reply:
This is guaranteed to be >= 0 which us what we want.

Reply:
`expr % 0` will cause a divide by zero error, which is probably not what we want.

Reply:
This is an Expr, so % 0 will give 0.